lib/functions/installs.bash

```diff
@@ -117,30 +121,146 @@ install_local_tool_versions() {
+
+  while [ "$search_path" != "/" ]; do
+    # install tools from files in current directory
+    display_debug "--------------------------------------------------------------------------------------------------------------"
+    tools_installed=$(install_directory_tools "$search_path" "$plugins_installed" "$tools_installed")
+    display_debug "install_directory_tools_recursive '$search_path': install_directory_tools returned tools_installed='$tools_installed'"
+
+    # terminate if all tools are installed
+    if [[ -n $(stringlist_same_length "$plugins_installed" "$tools_installed") ]]; then
+      display_debug "install_directory_tools_recursive '$search_path': exiting as tools_installed has same length as plugins_installed: tools_installed='$tools_installed', plugins_installed='$plugins_installed'"
+      printf "%s\n" "$tools_installed"
+      return 0
+    fi
+
+    # go up a directory
+    search_path=$(dirname "$search_path")
+  done
+
+  printf "%s\n" "$tools_installed"
+}
+
+install_directory_tools() {
+  local search_path=$1
+  local plugins_installed=$2
+  local tools_installed=$3
+  display_debug "install_directory_tools '$search_path': starting install. tools_installed='$tools_installed'"
+
+  # install tools from .tool-versions
+  # install order is the order listed in .tool-versions
+  local tool_versions
+  file_name=$(asdf_tool_versions_filename)
+  if ! [[ -f "$search_path/$file_name" ]]; then
+    display_debug "install_directory_tools '$search_path': exiting early... $file_name file not found"
+    printf "%s\n" "$tools_installed"
+    return 0
+  fi
+
+  tool_versions=$(strip_tool_version_comments "$search_path/$file_name" | awk '{$1=$1};1')
+  if [[ -z $tool_versions ]]; then
+    display_debug "install_directory_tools '$search_path': exiting early... no tools found in directory"
+    printf "%s\n" "$tools_installed"
+    return 0
+  fi
+  while IFS=' ' read -r tool_version; do
+    display_debug "install_directory_tools '$search_path': found '$tool_version'"
+
+    # read one version from the file
+    IFS=' ' read -ra parts <<< "$tool_version"
+    local plugin_name
+    plugin_name=${parts[0]}
+    local plugin_version
+    plugin_version=${parts[1]}
+
+    # skip if plugin is installed already
+    if [[ -n $(stringlist_contains "$tools_installed" "$plugin_name") ]]; then
+      display_debug "install_directory_tools '$search_path': '$plugin_name' is already installed... skipping"
+      continue
+    fi
+
+    # install the version
+    display_none $(install_tool_version "$plugin_name" "$plugin_version")
+    tools_installed=$(echo "$tools_installed $plugin_name" | awk '{$1=$1};1')
+
+    display_debug "install_directory_tools '$search_path': installed '$plugin_name':'$plugin_version' new state of tools_installed='$tools_installed'"
+  done <<< $tool_versions
+
+  # TODO when should we resolve tool version from the environment ?
```

Comment:
I believe you should be able to resolve the tool version from the environment... yet to add this feature